### PR TITLE
Fix gevent segfault with C-Ares as DNS resolver

### LINKED ISSUE #18422: gevent patch broken from v1.18.0 to v1.19.0
### What version of gRPC and what language are you using?

Python 3.7 

**grpcio==1.19.0** (Note: this started in 1.19.0 and is still broken in 1.19.1, it worked fine with 1.12.1 through 1.18.0)  

gevent==1.4.0
greenlet==0.4.15 ; platform_python_implementation == 'CPython'
gunicorn==19.9.0
 
### What operating system (Linux, Windows,...) and version?
 
Linux (AppEngine Flex VM)
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
 
Python 3.7
 
### What did you do?

We run a `flask` application on app engine flex, using `gunicorn` with `gevent` workers. We are using grpc via the http://github.com/googleapis/google-coud-python clients to talk to pubsub, cloud tasks, and datastore. Our app also does a bunch of other things (writes to cloud sql using sqlalchemy, uses `requests` to send http requests to other sites, etc.).

Our gevent worker is setup to call `init_gevent`:
```python
from gunicorn.workers.ggevent import GeventWorker
from grpc.experimental import gevent


class GeventGrpcWorker(GeventWorker):
    def patch(self):
        super(GeventGrpcWorker, self).patch()
        gevent.init_gevent()
        self.log.info('patched grpc')
```

and here is our gunicorn configuration:
```python
import multiprocessing
from gevent_grpc_worker import GeventGrpcWorker

# http://docs.gunicorn.org/en/stable/design.html#how-many-workers
workers = multiprocessing.cpu_count() * 2 + 1
worker_connections = 10000
# Use an asynchronous worker as most of the work is waiting for websites to load
worker_class = '.'.join([GeventGrpcWorker.__module__,
                         GeventGrpcWorker.__name__])
timeout = 30
```
 
### What did you expect to see?
 
We expected our workers to come up and start accepting requests
 
### What did you see instead?
 
Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

We receive segfaults and workers doesn't seem to be accepting requests:
```
Mar 18 21:43:19 aef-default-colinadams--gevent--v1--19--0-1p2z kernel: [  214.855792] gunicorn[4840]: segfault at 10 ip 00007fa121f9bf1a sp 00007ffdebdbc730 error 4 in cygrpc.cpython-37m-x86_64-linux-gnu.so[7fa121eae000+3af000]
```

I'll try to get more helpful debug information as well.

See [TROUBLESHOOTING.md](https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md) for how to diagnose problems better.
 
### Anything else we should know about your project / environment?

Seems similar to #17414 but for different versions. I can try paring this down to a minimal app that sees the issues.

### PR DESCRIPTION
**Pull request by lidizheng**
Fixes #18422

### CODE REVIEW

**Comment by markdroth** [comment]
I'll let @apolcyn review this, but from a quick glance, this doesn't look to me like the right solution.  If the goal here is to not use the c-ares resolver when gevent is enabled, then (a) I'd like to understand why c-ares won't work in that environment, and (b) it seems like we should find a way to disable it at compile-time instead of at run-time.

**Comment by lidizheng** [comment]
@markdroth The reason of c-ares segfault is that the `grpc_ares_ev_driver_posix.cc` depends on `g_event_engine`. Normally, the `g_event_engine` will be initialized during `grpc_init`, but the customized IO manager don't initialize it. So, the `NULL` pointer will trigger segfault when user makes a DNS query. This error will apply to Node.js if they enable `ares` at compile time.

For Python, currently we only ship one sets of binary wheels that users can switch their IO manager to `gevent` if needed. Either we disable `ares` at compile time for both native gRPC Python and `gevent` gRPC Python, or we have to ship the `gevent` gRPC Python as another PyPI package.

There are 3 options to fix this issue:
1. Build a new `grpc_ares_ev_driver` that support `gevent` as IO manager which could take a lot of time.
2. Set environment variable which might have strange side-effect.
3. Set a global variable to prevent using custom IO manager and C-Ares at the same time.

**Comment by apolcyn** [comment]
I agree with the comment from @lidizheng - the bug where g-event is broken on 1.19 was really somewhat of an oversight: the posix-specific code in the c-ares resolver uses the `grpc_fd_create`, `grpc_fd_orphan`, etc. APIs, and these APIs all assume that grpc is using the default posix iomgr, otherwise they just crash.

As mentioned, the way to make g-event work under c-ares would probably have to be by making some g-event specific code in the c-ares resolver, which knows how to create/read/write fd's under g-event, but that will take some time.

Regarding a compile-time fix, I'd note that compile-time feature enablement is problematic for all wrapped languages, since users tend to all use the same pre-compiled binary that we ship - requiring that users pre-compile from source to enable g-event would be an option, but would add a big hurdle to what people need to do today.

IMO we shouldn't be setting environment variables within the library, so I think this leaves the proposed fix as the best fix.

File: src/python/grpcio_tests/tests/unit/_dns_resolver_test.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class DNSResolverTest(unittest.TestCase):
+
+    def test_connect_loopback(self):
+        # NOTE(https://github.com/grpc/grpc/issues/18422)
+        # In short, Gevent + C-Ares = Segfault. The C-Ares driver is not
+        # supported by custom io manager like "gevent" or "libuv". This test is
+        # considered pass if the default DNS resolver didn't trigger segfault.
+        grpc.insecure_channel('loopback6.unittest.grpc.io')
```
Comment (apolcyn) [inline src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: Does this test fail without the fix?

Because we never make a call on this channel, and because we never try to explicitly connect, IIC this test won't attempt start connecting or resolving.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc [hunk 2]
```diff
@@ -433,8 +434,9 @@ static grpc_address_resolver_vtable ares_resolver = {
     grpc_resolve_address_ares, blocking_resolve_address_ares};
 
 static bool should_use_ares(const char* resolver_env) {
-  return resolver_env == nullptr || strlen(resolver_env) == 0 ||
-         gpr_stricmp(resolver_env, "ares") == 0;
+  return !g_custom_iomgr_enabled &&
```
Comment (apolcyn) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc]: Can we add a `TODO` comment that this can `g_custom_iomgr_enabled` check can go away after supporting c-ares on gevent (since the only other custom iomgr user, libuv, compiles out c-ares).

File: src/python/grpcio_tests/tests/unit/_dns_resolver_test.py (hunk 1 above)
Comment (lidizheng) [reply src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: Thank you for pointing out. I over-simplified the unit test that trying avoid the missing `IPv6` support in Kokoro (or Docker). But I found the existence of `loopback4`, and it will trigger the segfault in the same way. So, I revert to the original unit test, and use `IPv4` to connect.

**Comment by markdroth** [comment]
Before deciding how to handle this for gevent, we should talk with @guantaol about what the plan is for how to integrate gevent and the OSS EventManager API, since we're planning to basically replace most of the current iomgr API with the EventManager API.  Ideally, I would like to see this completely abstract away platform-specific behaviors so that c-ares doesn't need to be compiled with different code based on the platform.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc (hunk 2 above)
Comment (lidizheng) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc]: `TODO` comment added:

> // TODO(lidiz): Remove the "g_custom_iomgr_enabled" flag once c-ares support
>  // custom IO managers (e.g. gevent).

**Comment by apolcyn** [comment]
As I see, gevent basically needs to use the `grpc_custom_iomgr_init` API. Potential removal of that I think would also concern libuv, cc @murgatroid99

File: src/python/grpcio_tests/tests/unit/_dns_resolver_test.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    def tearDown(self):
+        self._server.stop(None)
+
+    def test_connect_loopback(self):
+        # NOTE(https://github.com/grpc/grpc/issues/18422)
+        # In short, Gevent + C-Ares = Segfault. The C-Ares driver is not
+        # supported by custom io manager like "gevent" or "libuv".
+        with grpc.insecure_channel(
+                'loopback4.unittest.grpc.io:%d' % self._port) as channel:
```
Comment (apolcyn) [inline src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: ipv6 should be available even in tests that run in docker containers on Kokoro.

Was the ipv6 availability issue encountered while running this test manually?

Comment (apolcyn) [reply src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: A bit more on that, IMO tests should assume that ipv6 is available, but not that ipv4 is available, because all grpc unit tests run on either dual stack or ipv6-only environments. Tests depending explicitly on ipv4 prevents such tests from ever running in ipv6-only environments, and futher keeps around the need for dual-stack.

Comment (lidizheng) [reply src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: The `IPv6` fail log is caused by the second commit, here is the full log: https://source.cloud.google.com/results/invocations/3799a132-06ea-494c-91c4-6466f18fa304/targets/grpc%2Fcore%2Fpull_request%2Flinux%2Fgrpc_python_bazel_test/log

Here is the highlight:
```
======================================================================
ERROR: test_connect_loopback (__main__.DNSResolverTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.cache/bazel/_bazel_root/d2dc70c3d9da3fab488ba0dcbbd35051/execroot/com_github_grpc_grpc/bazel-out/k8-fastbuild/bin/src/python/grpcio_tests/tests/unit/_dns_resolver_test.runfiles/com_github_grpc_grpc/src/python/grpcio_tests/tests/unit/_dns_resolver_test.py", line 53, in test_connect_loopback
    self.assertEqual(channel.unary_unary(_METHOD)(_REQUEST), _RESPONSE)
  File "/root/.cache/bazel/_bazel_root/d2dc70c3d9da3fab488ba0dcbbd35051/execroot/com_github_grpc_grpc/bazel-out/k8-fastbuild/bin/src/python/grpcio_tests/tests/unit/_dns_resolver_test.runfiles/com_github_grpc_grpc/src/python/grpcio/grpc/_channel.py", line 562, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/root/.cache/bazel/_bazel_root/d2dc70c3d9da3fab488ba0dcbbd35051/execroot/com_github_grpc_grpc/bazel-out/k8-fastbuild/bin/src/python/grpcio_tests/tests/unit/_dns_resolver_test.runfiles/com_github_grpc_grpc/src/python/grpcio/grpc/_channel.py", line 466, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
_Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "DNS resolution failed"
	debug_error_string = "{"created":"@1553714241.572645062","description":"Failed to create subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":2715,"referenced_errors":[{"created":"@1553714241.572633454","description":"Resolver transient failure","file":"src/core/ext/filters/client_channel/resolving_lb_policy.cc","file_line":335,"referenced_errors":[{"created":"@1553714241.572629488","description":"DNS resolution failed","file":"src/core/ext/filters/client_channel/resolver/dns/c_ares/dns_resolver_ares.cc","file_line":327,"grpc_status":14,"referenced_errors":[{"created":"@1553714241.572531047","description":"C-ares status is not ARES_SUCCESS: DNS server returned answer with no data","file":"src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc","file_line":240}]}]}]}"
>
```

What do you think is the best solution to satisfy both  Bazel in Kokoro/Docker and IPv6 only environment?

Comment (apolcyn) [reply src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: Hmm, yeah I guess these tests won't be able to use ipv6. In that case, use of the ipv4 loopback is<i>ok</i>

FWIW I checked the kokoro configs and it looks like the problem is that, unlike tests that use `build_docker_and_run_tests.sh` to invoke `docker run`, that bazel tests uses `build_and_run_docker.sh` to invoke `docker run` - and the latter doesn't supply the `--sysctl net.ipv6.conf.all.disable_ipv6=0` flag. It would probably be best to update `build_docker_and_run_tests.sh` to support ipv6 but I think that doesn't need to be done in this PR - cc @jtattermusch

File: src/python/grpcio_tests/tests/unit/_dns_resolver_test.py [hunk 4]
```diff
@@ -50,7 +51,12 @@ def test_connect_loopback(self):
         # supported by custom io manager like "gevent" or "libuv".
         with grpc.insecure_channel(
                 'loopback4.unittest.grpc.io:%d' % self._port) as channel:
-            self.assertEqual(channel.unary_unary(_METHOD)(_REQUEST), _RESPONSE)
+            response = channel.unary_unary(_METHOD)(
+                _REQUEST,
+                timeout=test_constants.SHORT_TIMEOUT,
+                wait_for_ready=True,
```
Comment (apolcyn) [inline src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: Why is this needed? Is this is needed to prevent flakiness in the test?

Comment (lidizheng) [reply src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: The `gevent` on Windows failed in last commit, and I am not 100% sure about the root cause so I use `wait-for-ready` to de-flake.

Test log highlight:
```
Traceback (most recent call last):
  File "C:\Python35\Lib\unittest\case.py", line 59, in testPartExecutor
    yield
  File "C:\Python35\Lib\unittest\case.py", line 601, in run
    testMethod()
  File "T:\src\github\grpc\workspace_python_windows_opt_gevent\src\python\grpcio_tests\tests\unit\_dns_resolver_test.py", line 53, in test_connect_loopback
    self.assertEqual(channel.unary_unary(_METHOD)(_REQUEST), _RESPONSE)
  File "T:\src\github\grpc\workspace_python_windows_opt_gevent\py35_gevent\lib\site-packages\grpc\_channel.py", line 562, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "T:\src\github\grpc\workspace_python_windows_opt_gevent\py35_gevent\lib\site-packages\grpc\_channel.py", line 466, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAVAILABLE
	details = "Connect Failed"
	debug_error_string = "{"created":"@1553726691.297000000","description":"Failed to create subchannel","file":"src/core/ext/filters/client_channel/client_channel.cc","file_line":2715,"referenced_errors":[{"created":"@1553726691.297000000","description":"failed to connect to all addresses","file":"src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc","file_line":444,"referenced_errors":[{"created":"@1553726691.297000000","description":"Connect Failed","file":"src/core/ext/filters/client_channel/subchannel.cc","file_line":963,"grpc_status":14,"referenced_errors":[{"created":"@1553726691.297000000","description":"connect failed: [Errno 10061] [WinError 10061] No connection could be made because the target machine actively refused it.","file":".\src/core/lib/iomgr/gevent_util.h","file_line":33,"grpc_status":14}]}]}]}"
>
```

Comment (apolcyn) [reply src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: One issue I think is that server is binding to `[::]` (the ipv6 loopback interface), but the client is trying to connect to the ipv4 loopback. I <i>think</i> that changing the server to bind to 127.0.0.1 would fix the issue.

I'm just worried about "wait-for-ready" masking underlying bugs....

Comment (lidizheng) [reply src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: Most of our Python servers are binding to `[::]` and the clients are connected to `localhost`. If the platform doesn't support IPv6, I presume the `localhost` will be resolved into `127.0.0.1`, IMO this test case's setup align with other test cases.

`wait-for-ready` flag should not affect correctness, but I agree that the connection failed after the start of server is definitely weird. I want to see if the problem go away with `gevent` Windows tests.

Comment (apolcyn) [reply src/python/grpcio_tests/tests/unit/_dns_resolver_test.py]: > `wait-for-ready` flag should not affect correctness

There are risks with using `wait-for-ready` for a test like this. If there is, for example, a bug in the resolver that causes only 1/10 DNS resolution attempts to succeed, then use of `wait-for-ready` would mask that issue and this test would always pass.... I think it's best if it can be avoided

**Comment by lidizheng** [comment]
After a week, I still don't understand why the test case failed in Windows. Here is the C-Core trace:
```
I0402 21:17:18.890000000  3400 src/core/ext/filters/client_channel/subchannel.cc:960] Connect failed: {"created":"@1554236238.890000000","description":"connect failed: [Errno 10061] [WinError 10061] No connection could be made because the target machine actively refused it.","file":".\src/core/lib/iomgr/gevent_util.h","file_line":33,"grpc_status":14}
I0402 21:17:18.890000000  3400 src/core/lib/transport/connectivity_state.cc:164] SET: 00000000019EF7E8 subchannel: CONNECTING --> TRANSIENT_FAILURE [connect_failed] error=0000000000F8B170 {"created":"@1554236238.890000000","description":"Connect Failed","file":"src/core/ext/filters/client_channel/subchannel.cc","file_line":963,"grpc_status":14,"referenced_errors":[{"created":"@1554236238.890000000","description":"connect failed: [Errno 10061] [WinError 10061] No connection could be made because the target machine actively refused it.","file":".\src/core/lib/iomgr/gevent_util.h","file_line":33,"grpc_status":14}]}
```

**Comment by lidizheng** [comment]
Known failure: https://github.com/grpc/grpc/issues/18523 https://github.com/grpc/grpc/issues/18608 https://github.com/grpc/grpc/issues/15411
